src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -89,38 +32,85 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn need_type_info_err(&self,
-                            body_id: Option<hir::BodyId>,
-                            span: Span,
-                            ty: Ty<'tcx>)
-                            -> DiagnosticBuilder<'gcx> {
+                              cause: &traits::ObligationCause<'tcx>,
+                              ty: Ty<'tcx>)
+                              -> DiagnosticBuilder<'gcx> {
         let ty = self.resolve_type_vars_if_possible(&ty);
         let name = self.extract_type_name(&ty);
 
-        let mut err_span = span;
         let mut labels = vec![(
-            span,
+            cause.span,
             if &name == "_" {
                 "cannot infer type".to_string()
             } else {
                 format!("cannot infer type for `{}`", name)
             },
         )];
+        let mut span = cause.span;
 
-        let mut local_visitor = FindLocalByTypeVisitor {
-            infcx: &self,
-            target_ty: &ty,
-            hir_map: &self.tcx.hir,
-            found_local_pattern: None,
-            found_arg_pattern: None,
-        };
-
-        if let Some(body_id) = body_id {
-            let expr = self.tcx.hir.expect_expr(body_id.node_id);
-            local_visitor.visit_expr(expr);
+        // NB. Lower values are more preferred.
+        #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+        enum LocalKind {
+            ClosureArg,
+            Let,
         }
 
-        if let Some(pattern) = local_visitor.found_arg_pattern {
-            err_span = pattern.span;
+        let found_local = self.in_progress_tables.and_then(|tables| {
+            let tables = tables.borrow();
+            let local_id_root = tables.local_id_root?;
+            assert!(local_id_root.is_local());
+
+            tables.node_types().iter().filter_map(|(&local_id, &node_ty)| {
```

Comment:
such ECS.

Reply:
Nit: It'd be nice to pull this out into some sort of helper function, I think? I'd also like to see a comment, something like:

---

Iterate over the all the things that we've assigned types to. Check whether they represent local variables and, if so, if the type of that local variable references the type we are looking for.